selfdrive/updated.py

```diff
@@ -302,11 +311,35 @@ def attempt_update():
 
     finalize_from_ovfs_copy()
 
+    # If a NEOS update is required, download it in the background
+    with open("/VERSION", "r") as current_neos_file:
+      current_neos_version = current_neos_file.read().replace("\n", "")
+    required_neos_version = run(["bash", "-c", r"source launch_env.sh && echo -n $REQUIRED_NEOS_VERSION"], FINALIZED)
+    if current_neos_version != required_neos_version or DEBUG_FORCE_UPDATE:
+      print(f"Beginning background download for NEOS {required_neos_version}")
+      try:
+        shutil.rmtree("/data/neoupdate")
+      except FileNotFoundError:
+        pass
+      update_json = f'file:///{FINALIZED}/installer/updater/update.json'
+      while True:
+        set_neos_download_param(True)
+        run(NICE_LOW_PRIORITY + ["installer/updater/updater", "bgcache", update_json], FINALIZED)
+        set_neos_download_param(False)
+        if os.path.isfile("/data/neoupdate/cache_success_marker"):
```

Comment:
I'm not sure I understand the point of the cache success marker. Can't you just check if the update file exists then check the hash?

Reply:
Diff minimization. The existing code rather freely mixes functional tasks and UI flows and updates in the same functions; I would have had to break it into pieces that can run headless and be called under varying circumstances. This method made maximal reuse of the existing tested paths.

Reply:
Seems like it would only take a couple more lines, no? You're already passing the update file to the updater which has the hash and check the actual update file instead of this cache success file?

Reply:
Per our discussion, I've moved things around to do without the cache semaphore file.